Web Inspector: Event badge popover remains visible after clicking to jump to event breakpoint

rcaliman-apple opened the pull request:
#### 17788c3b60cf8d86de1adea98b589664ac3e1049
<pre>
Web Inspector: Event badge popover remains visible after clicking to jump to event breakpoint
<a href="https://bugs.webkit.org/show_bug.cgi?id=255795">https://bugs.webkit.org/show_bug.cgi?id=255795</a>

Reviewed by Patrick Angle and Devin Rousso.

Make `WI.Popover` auto-dismiss whenever the selected tab of the `WI.TabBrowser` changes.

A popover is usually shown inside a tab in the tab browser. If the selected tab changes,
the context changes, so it makes sense to automatically dismiss the popover.

* Source/WebInspectorUI/UserInterface/Views/Popover.js:
(WI.Popover.prototype.dismiss):
(WI.Popover.prototype._addListenersIfNeeded):
(WI.Popover):

Canonical link: <a href="https://commits.webkit.org/272140@main">https://commits.webkit.org/272140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1e61db98812cdd127ee833ab1285f9cf449cef9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30689 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9369 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33187 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27744 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11669 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6626 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27632 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27478 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6739 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6901 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27347 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34524 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27940 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27833 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33058 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6923 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5023 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30885 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8646 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7260 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7647 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7476 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.js:
```diff
@@ -86,6 +86,9 @@ WI.DOMTreeOutline = class DOMTreeOutline extends WI.TreeOutline
     {
         this._visible = visible;
+
+        this.dispatchEventToListeners(WI.DOMTreeOutline.Event.VisibilityChanged, {visible});
```
i wonder if instead of just listening for this specific case we could/should instead listen for something more generic like `WI.ContentBrowser.Event.CurrentContentViewDidChange` and/or `WI.ContentViewContainer.Event.CurrentContentViewDidChange` 🤔

i wouldn't be surprised if there's other things like this elsewhere, so that would fix all of them (and future ones) in one go

rcaliman-apple replied:
> i wonder if instead of just listening for this specific case we could/should instead listen for something more generic

Do you mean adding something like this to `WI.Popover`?

```js
WI.tabBrowser.contentViewContainer.addEventListener("WI.ContentViewContainer.Event.CurrentContentViewDidChange", () => { 
    this.dismiss(); 
}, this);
```

(Assume the `contentViewContainer` getter exists on `WI.TabBrowser`)

This would solve the issue for all `WI.Popover` instances and make them self-dismiss.
But it makes `WI.Popover`  know about `WI.tabBrowser`. Is that ok?

rcaliman-apple replied:
> hmmm oddly i dont see this behavior when clicking on a source code location link in the WI.Popover 🤔
> and even weirder, once i've done that then i cant reproduce this issue

Indeed, clicking the link to jump to source doesn't exhibit the issue.
And yes, once that steps is done, jumping to the breakpoint also doesn't reproduce the issue anymore.

I thought it had something to do with this failed assertion. 
But the assertion is tripped in both scenarios: jump-to-breakpoint _and_  jump-to-source, then jump-to-breakpoint 🤔 

<img width="923" alt="Screenshot 2023-04-22 at 14 21 37" src="https://user-images.githubusercontent.com/105809798/233784502-af71d6fb-38ab-4ef8-b44b-bb720958d5f5.png">

dcrousso replied:
id love for us to understand why this is different and maybe integrate that with our solution

dcrousso reviewed (commented):
hmmm oddly i dont see this behavior when clicking on a source code location link in the `WI.Popover` 🤔

and even weirder, once i've done that then i cant reproduce this issue

i wonder why that scenario/codepath doesnt have this problem?  perhaps it has something to do with focusing/selecting in the shown `CodeMirror`?

patrickangle reviewed (approved):
I'm glad we have a solution that solves this more broadly than special-casing each popover to do this.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Popover.js:
```diff
@@ -598,6 +599,7 @@ WI.Popover = class Popover extends WI.Object
             window.addEventListener("scroll", this, true);
             window.addEventListener("resize", this, true);
             window.addEventListener("keypress", this, true);
+            WI.tabBrowser.addEventListener(WI.TabBrowser.Event.SelectedTabContentViewDidChange, this.dismiss, this);
```
so this is just specific to a tab changing.  what about if the current `WI.ContentView` with the same tab changes?  that's why id suggested `WI.ContentBrowser.Event.CurrentContentViewDidChange` or `WI.ContentViewContainer.Event.CurrentContentViewDidChange` because i think those will be even more broad and therefore probably be more safe (tho we'd probably want to verify that no `WI.Popover` creates a `WI.ContentBrowser` or `WI.ContentViewContainer` just in case)

rcaliman-apple committed:
Web Inspector: Event badge popover remains visible after clicking to jump to event breakpoint
https://bugs.webkit.org/show_bug.cgi?id=255795

Reviewed by Patrick Angle and Devin Rousso.

Make `WI.Popover` auto-dismiss whenever the selected tab of the `WI.TabBrowser` changes.

A popover is usually shown inside a tab in the tab browser. If the selected tab changes,
the context changes, so it makes sense to automatically dismiss the popover.

* Source/WebInspectorUI/UserInterface/Views/Popover.js:
(WI.Popover.prototype.dismiss):
(WI.Popover.prototype._addListenersIfNeeded):
(WI.Popover):

Canonical link: https://commits.webkit.org/272140@main
